strawberryfields/utils/post_processing.py

```diff
@@ -0,0 +1,271 @@
+# Copyright 2019-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module provides functions for post-processing samples.
+"""
+import numpy as np
+
+
+def _product_for_modes(samples, modes=None):
+    """Getting the product of samples across modes.
+Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): a flat sequence containing indices of modes to get
+            the expectation value for
+
+    Returns:
+        ndarray: product of the samples across modes
+    """
+    modes = np.array(modes)
+    selected_samples = samples[:, modes]
+    return np.prod(selected_samples, axis=1)
+
+
+def _check_samples(samples):
+    """Validation check for the input samples.
+
+    Checks include data types checks and dimension of the input samples.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape of (shots,
+            modes)
+    """
+    if not isinstance(samples, np.ndarray) or samples.ndim != 2:
+        raise Exception("Samples needs to be represented as a two dimensional NumPy array.")
+
+
+def _check_modes(samples, modes):
+    """Validation checks for the input modes.
+
+    Checks include data types checks, checks for the dimension of the inputs
+    and the validity of the modes specified.
+
+    Args:
+        samples (ndarray): the photon number samples with a shape
+            of (shots, modes)
+        modes (Sequence): the input modes to get the expectation value for, a
+            flattened sequence
+    """
+    num_modes = samples.shape[1]
+    flattened_sequence_indices_msg = (
+        "The input modes need to be specified as a flattened sequence of non-negative integers!"
+    )
+
+    modes = np.array(modes)
+
+    # Checking if modes is a valid flattened sequence of indices
+    try:
+        non_index_modes = modes[(np.mod(modes, 1) != 0) | (modes < 0)]
+    except Exception:
+        raise Exception(flattened_sequence_indices_msg)
+
+    if modes.ndim != 1 or non_index_modes.size > 0 or len(modes) == 0:
+        raise Exception(flattened_sequence_indices_msg)
+
+    # Checking if valid modes were specified
+    largest_valid_index = num_modes - 1
+    out_of_bounds_modes = modes[modes > largest_valid_index]
+    if out_of_bounds_modes.size > 0:
```

Comment:
Maybe clearer to just have the following?
```python
if max(modes) > num_modes - 1:
```

Reply:
thanks, that's a briefer and simpler way! :blush:

Reply:
oh just realized that the original solution also returns the invalid modes, so I'll leave it as is if it's okay